### PR TITLE
feat(spec): declared ### Skills section with fail-closed compile resolution

### LINKED ISSUE #60: Spec proposal: declared agent skills on components
# Spec proposal: declared agent skills on components

Filing this as a spec proposal first per maintainer guidance — no implementation,
just use case and proposed syntax for debate.

## Use case

OpenProse runs `.prose.md` components inside an agent harness (Claude Code, Codex,
etc.) that brings its own tools and skills. Today, a component cannot tell the
runtime which **harness skills** it depends on. That has two costs:

1. **Silent drift.** A component authored against `document-skills:pdf` will run
   on a machine where that skill is absent and the agent will quietly improvise —
   sometimes succeeding, sometimes producing wrong output, with no diagnostic.
2. **Unreviewable contracts.** A reviewer reading a `.prose.md` cannot see at a
   glance which agent capabilities the component leans on, even though those are
   load-bearing.

Concrete example: an `invoice-extractor` system whose `### Execution` calls a
PDF-parsing skill works on the author's box and fails on a teammate's box because
`document-skills:pdf` is not installed. The author wants to **declare** the
dependency so the runtime can fail closed at preflight with a named, actionable
diagnostic.

This is the same shape as `### Requires` / `### Ensures` for typed I/O ports —
just for harness capabilities instead of data flow.

## Ideal syntax

Two equivalent forms, both additive to the existing spec.

### Frontmatter form

```yaml
---
name: invoice-extractor
kind: system
skills:
  - document-skills:pdf
---
```

### Section form

```markdown
### Skills

- document-skills:pdf
- document-skills:xlsx
```

The two are merged (deduped by declared name) into the component's `skills`
field on the IR. Either alone is sufficient; using both is allowed and useful
when a sub-service adds a skill the parent system did not need.

### Naming

Skill names use the **colon form** that already matches plugin marketplace
conventions and `/skill` invocations:

```
namespace:name        # e.g. document-skills:pdf
```

A bare leaf (`pdf`) is accepted with a deterministic Levenshtein fuzzy match
against installed skills, and emits an informational diagnostic nudging the
author to pin the canonical name. Never silent.

### Scope

- **System-level** declarations apply to every sub-service inside the system.
- **Service-level** declarations are *additive*, not exclusive — declaring
  `skills:` on a `## sub-service` does not remove the system-level skills, it
  adds to them.

### BYO harness invariant

OpenProse never installs, modifies, or removes harness skills. BYO harness is
sacred. Preflight only **verifies** that declared skills are present and fails
closed (`skill_unresolved`) when one is missing, naming the skill and the search
paths it looked in. Installing the skill is the user's responsibility.

Search order:
1. The project's `./skills/` directory.
2. `~/.claude/skills/`.
3. `~/.codex/skills/`.

Resolved canonical names are pinned into the IR so subsequent runs of the same
IR are reproducible across machines.

## Where it slots into the existing spec

This is purely additive to `skills/open-prose/contract-markdown.md`:

- One new row in the **Frontmatter** table: `skills` (optional).
- One new row in the **Canonical Sections** table: `### Skills` (service, system).
- A short "Skill Declaration" subsection covering colon form, fuzzy fallback,
  scope, and the BYO invariant.

No existing sections change. No runtime semantics change for components that
don't declare skills.

## Alternatives considered

1. **Implicit detection** — scan `### Execution` for skill mentions. Rejected:
   too magical, and brittle against rephrasing or templating.
2. **Out-of-band manifest** (e.g., a `prose.skills.json` next to the file).
   Rejected: separates the contract from the dependency, defeating the
   "single reviewable file" property of `.prose.md`.
3. **Reuse `### Runtime`** for skills. Rejected: `Runtime` is for VM/model/thinking
   hints; mixing harness capability declarations there would muddy a section that
   is already load-bearing.
4. **Hard-fail on bare leaf names** instead of fuzzy match. Defensible, but
   reduces ergonomics for the common case where there is exactly one installed
   skill matching the leaf. Diagnostic + pin-on-resolve seems like the better
   default; happy to flip this if the maintainer prefers strict.

## Open questions for debate

1. Is `skills:` the right key, or should it be `harness-skills:` to make the
   "this is about the agent harness, not OpenProse internals" framing explicit?
2. Should the section header be `### Skills` or `### Harness Skills`?
3. Is the frontmatter form alone sufficient, or is the section form worth it
   for prose-readability? (The section form lets authors write a short
   justification line under each skill.)
4. Should the search order include a user-configurable path list, or is
   `./skills/` + `~/.claude/skills/` + `~/.codex/skills/` enough?
5. Should `skill_unresolved` block `prose run` outright, or default to a warning
   with a `--strict` flag? My instinct is fail-closed by default.

## What this issue is asking for

Agreement (or counter-proposal) on:
- The use case
- The two-form syntax (frontmatter + section)
- The colon naming convention and fuzzy fallback policy
- The BYO harness invariant and search order

Once the spec is settled, I'll send a small doc-only PR adding just the spec
text to `skills/open-prose/contract-markdown.md` plus a single assertion test,
and we can debate implementation separately.

### PR DESCRIPTION
**Pull request by rawwerks**
Implements the spec from #60 with the maintainer's review applied verbatim.

## Maintainer's decisions from #60 (applied)

| Question | Decision | Where |
|---|---|---|
| `### Skills` canonical section | Yes | `skills/open-prose/contract-markdown.md` (Canonical Sections + Skills H2) |
| `skills:` frontmatter form | Dropped — section only, one place | n/a |
| Section header name | `### Skills` (not `### Harness Skills`) | spec |
| Search order | `./skills`, `~/.claude/skills`, `~/.codex/skills`, `~/.agents/skills` | `tools/cli/src/skills/declared.ts` (`DECLARED_SKILL_SEARCH_DIRECTORIES`) |
| `~/.agents/skills` added | Yes (everyone-except-claude convention) | same |
| Fail behavior | Fail closed on `prose compile` | `tools/cli/src/commands/compile.ts` (`preflightDeclaredSkillsForCompile`) |
| BYO harness invariant | Confirmed — OpenProse never installs skills | spec |
| Scope | Minimal. One example/test. | this PR |

## What changed (8 files, +823)

- **Spec** — `skills/open-prose/contract-markdown.md` gains a `### Skills` row in the Canonical Sections table and a short `## Skills` H2 covering colon naming, search order, the BYO invariant, and fail-closed semantics.
- **Implementation** — `tools/cli/src/skills/declared.ts` (new). Pure TS: `parseDeclaredSkills`, `resolveDeclaredSkill`, `resolveDeclaredSkillsForFile`, `preflightDeclaredSkillsInRoot`, `DeclaredSkillsUnresolvedError`. No I/O beyond `readFile` / `readdir` / `stat`.
- **Wiring** — `tools/cli/src/commands/compile.ts` pre-checks declared skills before forwarding the compile prompt; throws `CompileValidationError` with a `skill_unresolved` heading and per-skill detail lines listing the searched paths. Pre-forward, so unresolved skills do not burn agent tokens. Gated behind the existing `skillPreflight` option for test parity.
- **Example** — `skills/open-prose/examples/declared-skills/` shows the `document-skills:pdf` canonical pattern from the issue.
- **Tests** — 19 new (TDD):
  - 18 in `tools/cli/tests/skills/declared.test.ts` covering parser, resolver, directory walker, error formatter, and edge cases (case-insensitive header, dedupe, backticked names, hidden-dir skipping).
  - 1 in `tools/cli/tests/cli/cli.test.ts` asserting `prose compile` rejects with `CompileValidationError` *before* the harness is invoked when a declared skill is missing.
- **Changelog** — `[Unreleased]` entry references #60.

Total test count: 158 → 177 (all green, typecheck clean).

## Test plan

- [x] `npx vitest run` — 177 / 177 pass
- [x] `npx tsc -p tsconfig.json --noEmit` — clean
- [x] Manual: a `.prose.md` declaring `document-skills:pdf` against an empty HOME triggers `skill_unresolved` with searched paths, harness uninvoked

## Notes

Branched off `origin/main` this time (last attempt was anchored to the abandoned `rfc/reactive-openprose` branch — that PR (#59) is closed). No frontmatter form, per maintainer feedback that frontmatter stays structural. `prose run` enforcement was deliberately deferred — once `prose compile` is fail-closed, a successful compile guarantees declared skills resolved at compile time, and runtime enforcement can be a follow-up if needed.

Closes #60.

Co-Authored-By: Claude Opus 4.7 <noreply@anthropic.com>

### CODE REVIEW

File: skills/open-prose/contract-markdown.md [hunk 1]
```diff
@@ -392,6 +393,47 @@ Rules:
...
+- document-skills:xlsx
+```
+
+Skill names use the `namespace:name` colon form that matches the plugin
+marketplace convention shown in `/skill` invocations.
+
+Rules:
+
+- `### Skills` is the only place to declare harness-skill dependencies. The
+  declaration is not duplicated in frontmatter — frontmatter stays structural
```
Comment (irl-dan) [inline skills/open-prose/contract-markdown.md]: we can remove lines like this.

Common claude/codex failure pattern: I often see design decisions creep into implementation. They proposed adding it to the front-matter, we said no, and then they propagate a description of that decision into the actual implementation when it would be better to just not say it at all.

I see this all the time.

File: skills/open-prose/contract-markdown.md [hunk 2]
```diff
@@ -392,6 +393,47 @@ Rules:
...
+  Service-level declarations are *additive*, not exclusive — declaring `###
+  Skills` on a `## sub-service` does not remove the system-level skills, it
+  adds to them.
+- Forme resolves each declared skill by looking, in order, in:
+  1. The project's `./skills/` directory.
+  2. `~/.claude/skills/`.
+  3. `~/.codex/skills/`.
+  4. `~/.agents/skills/`.
+- If a declared skill cannot be resolved in any of those paths, both `prose
+  compile` and `prose run` fail closed with a `skill_unresolved` diagnostic
```
Comment (irl-dan) [inline skills/open-prose/contract-markdown.md]: Here it says `prose run` fails, but in the PR description it says we deferred that decision and that `prose run` will not fail.

File: tools/cli/src/skills/declared.ts [hunk 3]
```diff
@@ -0,0 +1,226 @@
+import { readdir, readFile, stat } from "node:fs/promises";
```
Comment (irl-dan) [inline tools/cli/src/skills/declared.ts:7]: I don't think we want the logic for "searching for skills that are required" to be a harness responsibility.

It should be a compiler/program-level responsibility.

Keep in mind the goal for `prose` is to still be harness-agnostic, and if we put this logic here instead of in the `compiler` (in markdown), then running a Prose-with-Skills program on a different harness will not behave as we expect.

File: skills/open-prose/contract-markdown.md (hunk 2 above)
Comment (irl-dan) [reply skills/open-prose/contract-markdown.md]: I think it's fine to do it just in the `prose compile` step, but I'm curious why it came to the decision to defer it for `prose run` to make sure I'm not missing anything.

**Comment by rawwerks** [comment]
Dan, yes, the review comments on #62 were addressed in `d4ff12a`:

- `contract-markdown.md`: removed the rejected frontmatter-alternative wording. The spec now just describes the accepted `### Skills` section without carrying forward the “not duplicated in frontmatter” decision text.
- `contract-markdown.md`: corrected the fail-closed wording to say `prose compile` only. `prose run` enforcement remains deferred, matching the PR description.
- `compiler/index.prose.md`: moved the behavior into the compiler/program contract via the `skills_resolver` agent. That agent owns the search order, scope aggregation, BYO invariant, and `skill_unresolved` diagnostics. The CLI TypeScript remains the current harness implementation of that compiler step, but the semantics now live in the compiler program so other harnesses can implement the same behavior.
